velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+class S3WriteFile final : public WriteFile {
+ public:
+  explicit S3WriteFile(const std::string& path, Aws::S3::S3Client* client)
+      : client_(client) {
+    bucketAndKeyFromS3Path(path, bucket_, key_);
+  }
+
+  void initialize() {
+    // Check that the object doesn't exist, if it does throw an error.
+    {
+      Aws::S3::Model::HeadObjectRequest request;
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+      auto objectMetadata = client_->HeadObject(request);
+      VELOX_CHECK(!objectMetadata.IsSuccess(), "File already exists");
+    }
+
+    // Create bucket if not present.
+    {
+      Aws::S3::Model::HeadBucketRequest request;
+      request.SetBucket(awsString(bucket_));
+      auto bucketMetadata = client_->HeadBucket(request);
+      if (!bucketMetadata.IsSuccess()) {
+        Aws::S3::Model::CreateBucketRequest request;
+        request.SetBucket(bucket_);
+        auto outcome = client_->CreateBucket(request);
+        VELOX_CHECK_AWS_OUTCOME(
+            outcome, "Failed to create S3 bucket", bucket_, "");
+      }
+    }
+
+    // Initiate the multi-part upload.
+    {
+      Aws::S3::Model::CreateMultipartUploadRequest request;
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+
+      /// If we do not set anything then the SDK will default to application/xml
+      /// which confuses some tools
+      /// (https://github.com/apache/arrow/issues/11934). So we instead default
+      /// to application/octet-stream which is less misleading.
+      request.SetContentType("application/octet-stream");
+
+      auto outcome = client_->CreateMultipartUpload(request);
+      VELOX_CHECK_AWS_OUTCOME(
+          outcome, "Failed initiating multiple part upload", bucket_, key_);
+      uploadState_.id = outcome.GetResult().GetUploadId();
+    }
+
+    // Create a StringStream to store the writes.
+    currentPart_ = Aws::MakeShared<Aws::StringStream>(
+        "PutObjectInputStream",
+        std::stringstream::in | std::stringstream::out |
+            std::stringstream::binary);
+    closed_ = false;
+    fileSize_ = 0;
+  }
+
+  // Appends data to the end of the file.
+  void append(std::string_view data) override {
+    VELOX_CHECK(!closed_, "File is closed");
+    // 'flush' API should trigger uploadPart.
+    // But upload part if the maximum part size is reached.
+    if (currentPartSize_ + data.size() > kMaxPartSize) {
+      uploadPart();
+      currentPartSize_ = 0;
+    }
+    currentPart_->write(data.data(), data.size());
+    currentPartSize_ += data.size();
+    fileSize_ += data.size();
+  }
+
+  // Upload the current part.
+  void flush() override {
+    VELOX_CHECK(!closed_, "File is closed");
+    // Skip flushing an empty part.
+    if (currentPartSize_ == 0) {
+      return;
+    }
+    // Skip if current part size is less than the minimum part size.
+    if (currentPartSize_ < kMinPartSize) {
+      return;
+    }
+    uploadPart();
```

Comment:
should this be within a loop, till all parts are flushed, or is that with the caller of the flush method?

Reply:
There is only a single part being populated via `append()` at a given time. When the caller invokes `flush`, the entire data buffered is uploaded as a single part. All the current and previous parts are on S3.